File: plugins/stuffbucket/skills/tauri-events/templates/events.ts
// Tauri v2 — events and channels (frontend side).
//
// Global events: subscribe with `listen`, always call the returned unlisten.
// Channels: pass a Channel<T> as a command argument, consume via .onmessage.

import { invoke, Channel } from '@tauri-apps/api/core';
import { listen, type UnlistenFn, type Event } from '@tauri-apps/api/event';

// ---------- Global event consumer ----------

interface StatusChanged {
  state: string;
  atMs: number;
}

export async function subscribeStatus(
  onChange: (s: StatusChanged) => void,
): Promise<UnlistenFn> {
  return listen<StatusChanged>('status-changed', (e: Event<StatusChanged>) => {
    onChange(e.payload);
  });
}

// Example usage with explicit teardown:
//
//   const unlisten = await subscribeStatus((s) => render(s));
//   // ...later:
//   unlisten();

// React-flavored pattern (drop into a component):
//
//   useEffect(() => {
//     let off: UnlistenFn | undefined;
//     subscribeStatus(setStatus).then((fn) => { off = fn; });
//     return () => { off?.(); };
//   }, []);

// ---------- Trigger broadcast / targeted emit from JS ----------

export function broadcastStatus(state: string): Promise<void> {
  return invoke('broadcast_status', { state });
}

export function notifyWindow(label: string, message: string): Promise<void> {
  return invoke('notify_window', { label, message });
}

// ---------- Channel-based streaming consumer ----------

export type ProgressEvent =
  | { event: 'started'; data: { total: number } }
  | { event: 'tick'; data: { done: number } }
  | { event: 'finished'; data: { ok: boolean } };

export interface ProgressHandlers {
  onStart?: (total: number) => void;
  onTick?: (done: number, total: number) => void;
  onDone?: (ok: boolean) => void;
}

export async function runTask(
  total: number,
  handlers: ProgressHandlers = {},
): Promise<void> {
  const onEvent = new Channel<ProgressEvent>();
  let lastTotal = total;

  onEvent.onmessage = (msg) => {
    switch (msg.event) {
      case 'started':
        lastTotal = msg.data.total;
        handlers.onStart?.(lastTotal);
        break;
      case 'tick':
        handlers.onTick?.(msg.data.done, lastTotal);
        break;
      case 'finished':
        handlers.onDone?.(msg.data.ok);
        break;
    }
  };

  // The channel stays open until this promise resolves and the channel is GC'd.
  await invoke('run_task', { total, onEvent });
}

// Example:
//
//   await runTask(100, {
//     onStart: (n) => console.log('starting', n),
//     onTick:  (d, t) => bar.set(d / t),
//     onDone:  (ok) => console.log('done', ok),
//   });

File: plugins/stuffbucket/skills/tauri-events/templates/events.rs
// Tauri v2 — events and channels reference handler.
//
// Wire into your builder:
//   tauri::Builder::default()
//     .invoke_handler(tauri::generate_handler![
//       broadcast_status,
//       notify_window,
//       run_task,
//     ])

use serde::Serialize;
use tauri::{ipc::Channel, AppHandle, Emitter, EventTarget, Listener};

// ---------- Global event ----------

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusChanged {
  pub state: String,
  pub at_ms: u64,
}

#[tauri::command]
pub fn broadcast_status(app: AppHandle, state: String) -> Result<(), String> {
  let payload = StatusChanged {
    state,
    at_ms: std::time::SystemTime::now()
      .duration_since(std::time::UNIX_EPOCH)
      .map(|d| d.as_millis() as u64)
      .unwrap_or(0),
  };
  // Broadcast to every webview that listens for "status-changed".
  app.emit("status-changed", payload).map_err(|e| e.to_string())
}

// ---------- Targeted emit ----------

#[tauri::command]
pub fn notify_window(app: AppHandle, label: String, message: String) -> Result<(), String> {
  // Only the webview with the given label receives this.
  app
    .emit_to(EventTarget::webview_window(&label), "notice", message)
    .map_err(|e| e.to_string())
}

// ---------- Channel streaming ----------

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase", tag = "event", content = "data")]
pub enum ProgressEvent {
  Started { total: u64 },
  Tick { done: u64 },
  Finished { ok: bool },
}

#[tauri::command]
pub async fn run_task(
  total: u64,
  on_event: Channel<ProgressEvent>,
) -> Result<(), String> {
  on_event
    .send(ProgressEvent::Started { total })
    .map_err(|e| e.to_string())?;

  for i in 1..=total {
    // Imagine real work here; channels handle high-frequency sends.
    on_event
      .send(ProgressEvent::Tick { done: i })
      .map_err(|e| e.to_string())?;
  }

  on_event
    .send(ProgressEvent::Finished { ok: true })
    .map_err(|e| e.to_string())?;
  Ok(())
}

// ---------- Rust-side listener (e.g. in setup) ----------

pub fn wire_internal_listeners(app: &AppHandle) {
  let id = app.listen("frontend-ready", |event| {
    println!("frontend says: {}", event.payload());
  });
  // Keep `id` if you need to call app.unlisten(id) later.
  let _ = id;

  app.once("first-paint", |_event| {
    println!("first paint observed");
  });
}
